### PR TITLE
Merge into empty master

### PR DESCRIPTION
**Pull request by MohamedAbdelsalam9**

### CODE REVIEW

File: experiments/main.py [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+    parser.add_argument('--node_rank', default=0, type=int,
+                        help='node rank for distributed training')
+    parser.add_argument('--multiprocessing_distributed', action='store_true',
+                        help='Use multi-processing distributed training to launch '
+                             'N processes per node, which has N GPUs.')
+    parser.add_argument('--dist_url', default="env://", type=str,
+                        help='node rank for distributed training')
+
+    args = parser.parse_args()
+    args.num_workers = 6
```
Comment (GabPrato) [inline experiments/main.py]: Is there a reason why this is not a hyper-parameter?

File: docs/build/html/_sources/lifelong_methods_guide.md.txt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    
+    # This is where anything that needs to be done after the task is done takes place
+    example_model.consolidate_task_knowledge(**kwargs)
+
+    # Inference
+    # This is where the inference happens
+    predictions = example_model(inference_data_batch)
+```
+
+This is the typical order of how things flow in a lifelong learning scenario, and how does this package handle
```
Comment (AbdelrahmanZayed) [inline docs/build/html/_sources/lifelong_methods_guide.md.txt:30]: very minor comment: "and how this package handles that"

File: experiments/main.py [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+
+    config = prepare_config(args)
+    if "iirc" in config["dataset"]:
+        config["setup"] = IIRC_SETUP
+    else:
+        config["setup"] = CIL_SETUP
+
+    config['device'] = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+    config["ngpus_per_node"] = torch.cuda.device_count()
```
Comment (GabPrato) [inline experiments/main.py:401]: This is not a general solution. Some people might run your code on setups where they have to use specific gpus, like cuda:3 and cuda:5. To support any possible configuration, you have to allow the user to specify with args which gpus to use. Not the end of the world, but keep in mind that some people might have to modify your code to be able to run it.

File: experiments/main.py [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+        else:
+            dist.barrier()
+            logging.basicConfig(filename=os.path.join(config['logging_path'], "logs.txt"),
+                                filemode='a',
+                                format='%(asctime)s,%(msecs)d %(name)s %(levelname)s %(message)s',
+                                datefmt='%H:%M:%S',
+                                level=logging.INFO)
+
+    torch.random.manual_seed(config['seed'])
+    np.random.seed(config['seed'])
```
Comment (GabPrato) [inline experiments/main.py:144]: If you are generating random numbers with numpy in your workers, they will all generate the same random numbers. If this is what you want, that is fine, if not, then be very careful with these things. E.g., if you are using workers to randomly sample from your dataset, all workers will fetch the same samples.

File: experiments/train.py [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+            labels_names, num_seen_classes, class_names_to_idx
+        )
+
+        if gpu is None:
+            images = minibatch[0].to(config["device"], non_blocking=True)
+            labels = labels.to(config["device"], non_blocking=True)
+        else:
+            images = minibatch[0].to(torch.device(f"cuda:{gpu}"), non_blocking=True)
+            labels = labels.to(torch.device(f"cuda:{gpu}"), non_blocking=True)
+
```
Comment (GabPrato) [inline experiments/train.py:36]: If you set config["device"] = torch.device('cpu') or 'cuda', then you don't need to do an if else.

**Review by GabPrato — APPROVED** [review]

File: experiments/train.py [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+        train_metrics['modified_jaccard'] += metrics.modified_jaccard_sim(predictions, labels) * images.shape[0]
+        train_metrics['strict_acc'] += metrics.strict_accuracy(predictions, labels) * images.shape[0]
+        train_metrics['recall'] += metrics.recall(predictions, labels) * images.shape[0]
+        data_len += images.shape[0]
+
+        if minibatch_i == 0:
+            print_msg(
+                f"rank {rank}, max memory allocated: {torch.cuda.max_memory_allocated() / 1024 / 1024} MB, "
+                f"current memory allocated: {torch.cuda.memory_allocated() / 1024 / 1024} MB\n")
+        # if minibatch_i % 100 == 0:
```
Comment (AbdelrahmanZayed) [inline experiments/train.py]: Is this commented part used in the code?

**Review by AbdelrahmanZayed — APPROVED** [review]

File: experiments/train.py (hunk 5 above)
Comment (MohamedAbdelsalam9) [reply experiments/train.py:36]: this is for the distributed computing when running on multiple gpu (which is not captured by config["device"])
